usr/sbin/p11sak/p11sak.c

```diff
@@ -869,45 +957,28 @@ static CK_RV tok_key_gen(CK_SESSION_HANDLE session, CK_ULONG keylength,
     CK_BBOOL a_token = ckb_true; // always true
     CK_BBOOL a_private = ckb_true; // always true
 
-    /* Boolean attributes from user input */
-    CK_BBOOL a_modifiable = ckb_false;
-    CK_BBOOL a_derive = ckb_false;
-    CK_BBOOL a_sensitive = ckb_false;
-    CK_BBOOL a_encrypt = ckb_false;
-    CK_BBOOL a_decrypt = ckb_false;
-    CK_BBOOL a_sign = ckb_false;
-    CK_BBOOL a_verify = ckb_false;
-    CK_BBOOL a_wrap = ckb_false;
-    CK_BBOOL a_unwrap = ckb_false;
-    CK_BBOOL a_extractable = ckb_false;
-    CK_ULONG bs = sizeof(CK_BBOOL);
-
     /* Non-boolean attributes */
     CK_ULONG a_value_len = keylength / 8;
 
-    CK_ATTRIBUTE tmplt[] = {
-            // boolean attrs
-            { CKA_TOKEN, &a_token, bs }, { CKA_PRIVATE, &a_private, bs }, {
-            CKA_MODIFIABLE, &a_modifiable, bs }, { CKA_DERIVE, &a_derive, bs },
-            { CKA_SENSITIVE, &a_sensitive, bs }, {
-            CKA_ENCRYPT, &a_encrypt, bs }, { CKA_DECRYPT, &a_decrypt, bs }, {
-                    CKA_SIGN, &a_sign, bs }, {
-            CKA_VERIFY, &a_verify, bs }, { CKA_WRAP, &a_wrap, bs }, {
-            CKA_UNWRAP, &a_unwrap, bs },
-            { CKA_EXTRACTABLE, &a_extractable, bs },
-            // non-boolean attrs
-            { CKA_VALUE_LEN, &a_value_len, sizeof(CK_ULONG) }, { CKA_LABEL,
-                    label, strlen(label) } };
-    CK_ULONG num_attrs = sizeof(tmplt) / sizeof(CK_ATTRIBUTE);
-    CK_ULONG num_bools = num_attrs - 2;
-
-    /* set boolean attributes */
-    for (i = 0; i < (int) num_bools; i++) {
-        set_bool_attr_from_string(&tmplt[i], attr_string);
+    /* Standard template */
+    CK_ATTRIBUTE key_attr[4 + KEY_MAX_BOOL_ATTR_COUNT] = {
+        { CKA_TOKEN, &a_token, sizeof(CK_BBOOL) }, 
+        { CKA_PRIVATE, &a_private, sizeof(CK_BBOOL) }, 
+        { CKA_VALUE_LEN, &a_value_len, sizeof(CK_ULONG) }, 
+        { CKA_LABEL, label, strlen(label) } 
+    };
+    CK_ULONG num_attrs = 4;
+    
+    /* set boolean attributes, set template from string */
+    if (attr_string) {
+        for (i = 0; i < (int) strlen(attr_string); i++) {
```

Comment:
Is there a check that `strlen(attr_string) <= KEY_MAX_BOOL_ATTR_COUNT`?  Otherwise you might have an overwrite of some arbitrary memory on the stack here and later further down in the read.

Reply:
The same applies to the `set_battr()` function above. The caller provides an attribute array and `set_battr()` adds attributes to the array based on the attribute string and may overrun the bounds of the caller supplied array if the attribute string contains more characters than the array slots.